app/src/main/java/com/example/contentprovider_example/ui/MainActivity.kt

```diff
@@ -0,0 +1,37 @@
+package com.example.contentprovider_example.ui
+
+import android.net.Uri
+import androidx.appcompat.app.AppCompatActivity
+import android.os.Bundle
+import androidx.recyclerview.widget.LinearLayoutManager
+import com.example.contentprovider_example.R
+import com.example.contentprovider_example.data.model.Employee
+import com.example.contentprovider_example.data.source.local.provider.EmployeeProvider
+import com.example.contentprovider_example.ui.adapter.EmployeeAdapter
+import kotlinx.android.synthetic.main.activity_main.*
+
+class MainActivity : AppCompatActivity(), MainContracts.View {
+
+    private lateinit var mainPresenter: MainPresenter
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+
+        mainPresenter = MainPresenter(this)
+        // use main activity as other application to get data from content provider
+        getDataFromProvider()
+    }
+
+    private fun getDataFromProvider() {
+        val uri = Uri.parse(EmployeeProvider.URI)
+        mainPresenter.handleInitializeData(contentResolver, uri)
+    }
+
+    override fun onDataInitializeSuccess(employeeList: ArrayList<Employee>) {
```

Comment:
View chỉ chứa những hàm tên gọi như showA, showB, updateA 
Logic của nó được xử lý dưới presenter nên không đặt tên hàm kiểu như success, fail, như vậy là biết logic rồi đó 
Đặt kiểu tham số rộng nhất có thể nhé 
List -> MutableList -> ArrayList 
```suggestion
    override fun showEmployees(employees: List<Employee>) {
```

Reply:
yep

Reply:
> View chỉ chứa những hàm tên gọi như showA, showB, updateA
> Logic của nó được xử lý dưới presenter nên không đặt tên hàm kiểu như success, fail, như vậy là biết logic rồi đó
> Đặt kiểu tham số rộng nhất có thể nhé
> List -> MutableList -> ArrayList

yep